### DISCUSSION TITLE
CARD tag in STMTTRN

### DISCUSSION THREAD

**Discussion by jslttery** [csingley/ofxtools#125] [General]
I tried to parse an OFX file from BarclayCard UK this evening and got a failure because the ofxtools library didn't recognise a CARD tag inside a STMTTRN tag - example:

      <STMTTRN>
        <TRNTYPE>DEBIT</TRNTYPE>
        <DTPOSTED>20200714000000</DTPOSTED>
        <TRNAMT>-9.99</TRNAMT>
        <FITID>20200714000000001</FITID>
        <NAME>Spotify UK, London</NAME>
        <CARD>Visa</CARD>
        <MEMO>1006/9.99 POUND STERLING GREAT BRITAIN</MEMO>
      </STMTTRN>

I've patched my local library as follows - in ofxtools/models/bank/stmt.py, edit class STMTTRN to put in CARD just before MEMO:

    ccacctto = SubAggregate(CCACCTTO)
    card = String(255)
    memo = String(255)

The parse appears to to be working now - I'll do some more tests.

If this is valid OFX, could the library be updated to incorporate this?

Thanks in advance,

S.

**Comment by csingley** [comment]
Thanks for reporting.  What version of OFX does your data report in the OFX header?

**Comment by jslttery** [comment]
Thanks for looking into this - here’s the header:

OFXHEADER:100
DATA:OFXSGML
VERSION:102
SECURITY:NONE
ENCODING:USASCII
CHARSET:1252
COMPRESSION:NONE
OLDFILEUID:NONE
NEWFILEUID:NONE

**Comment by csingley** [comment]
`STMTTRN` is defined in section 11.4.2.3.1 of the OFXv1.02 specification.  There is a distinct absence of any such `CARD` element therein defined.

Barclays, Barclays, Barclays.  If you want to define a proprietary OFX tag, it needs to be scoped to a namespace and delineated by a dot.

I'm afraid this is a fairly gross violation of the spec.  What's puzzling to me is why Quicken would parse that, or alternatively, how Barclays could manage to keep producing data downloads that can't be parsed by Quicken.

You should probably file a bug report to Barclays on general principles, but I wouldn't hold my breath waiting for a fix from them.

**Comment by jslttery** [comment]
Will see if I can find a way to flag this up to BarclayCard - but agree that they’re unlikely to correct it.

Will come back if I get anything useful back. For now I’m happy to use my locally hacked version of ofxtools.

Many thanks for a very useful library!

S.
